rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableMetaCachePostgresql.java

```diff
@@ -0,0 +1,200 @@
+        .expireAfterWrite(EXPIRE_TIME, TimeUnit.MILLISECONDS).softValues().build();
+
+    private static Logger logger = LoggerFactory.getLogger(TableMetaCachePostgresql.class);
+
+    /**
+     * Gets table meta.
+     *
+     * @param dataSourceProxy the druid data source
+     * @param tableName the table name
+     * @return the table meta
+     */
+    public static TableMeta getTableMeta(final DataSourceProxy dataSourceProxy, final String tableName) {
+        if (StringUtils.isEmpty(tableName)) {
+            throw new IllegalArgumentException("TableMeta cannot be fetched without tableName");
+        }
+
+        String dataSourceKey = dataSourceProxy.getResourceId();
+
+        TableMeta tmeta = null;
+        final String key = dataSourceKey + "." + tableName;
+        tmeta = TABLE_META_CACHE.get(key, mappingFunction -> {
+            try {
+                return fetchSchema(dataSourceProxy.getTargetDataSource(), tableName);
+            } catch (SQLException e) {
+                logger.error("get cache error !", e);
+                return null;
+            }
+        });
+        if (tmeta == null) {
+            try {
+                tmeta = fetchSchema(dataSourceProxy.getTargetDataSource(), tableName);
+            } catch (SQLException e) {
+            }
+        }
+
+        if (tmeta == null) {
+            throw new ShouldNeverHappenException(String.format("[xid:%s]get tablemeta failed", RootContext.getXID()));
+        }
+        return tmeta;
+    }
+
+    private static TableMeta fetchSchema(DataSource dataSource, String tableName) throws SQLException {
+        return fetchSchemeInDefaultWay(dataSource, tableName);
+    }
+
+    private static TableMeta fetchSchemeInDefaultWay(DataSource dataSource, String tableName)
+        throws SQLException {
+        Connection conn = null;
+        java.sql.Statement stmt = null;
+        try {
+            conn = dataSource.getConnection();
+            stmt = conn.createStatement();
+            DatabaseMetaData dbmd = conn.getMetaData();
+            return resultSetMetaToSchema(null, dbmd, tableName);
+        } catch (Exception e) {
+            if (e instanceof SQLException) {
+                throw ((SQLException) e);
+            }
+            throw new SQLException("Failed to fetch schema of " + tableName, e);
+
+        } finally {
+            if (stmt != null) {
+                stmt.close();
+            }
+        }
+    }
+
+    private static TableMeta resultSetMetaToSchema(ResultSetMetaData rsmd, DatabaseMetaData dbmd, String tableName)
+        throws SQLException {
+        TableMeta tm = new TableMeta();
+        tm.setTableName(tableName);
+        String[] schemaTable = tableName.split("\\.");
+        String schemaName = schemaTable.length > 1 ? schemaTable[0] : "public";
+        tableName = schemaTable.length > 1 ? schemaTable[1] : tableName;
+
+        ResultSet rs1 = dbmd.getColumns("", schemaName, tableName, "%");
```

Comment:
please close the rs1, rs2 and pk.
`
} finally {
            if (rsColumns != null) {
                rsColumns.close();
            }
            if (rsIndex != null) {
                rsIndex.close();
            }
            if (rsPrimary != null) {
                rsPrimary.close();
            }
}
`

Reply:
fix the conflict

Reply:
use try(ResultSet rsColumns = dbmd...)
ResultSet extends AutoCloseable